lib/service/service.go

```diff
@@ -3797,6 +3797,14 @@ func (process *TeleportProcess) singleProcessMode() (string, bool) {
 			return process.Config.Proxy.PublicAddrs[0].String(), true
 		}
 
+		// If WebAddress is unspecified "0.0.0.0" replace 0.0.0.0 with localhost since 0.0.0.0 is never a valid
+		// principal (auth server explicitly removes it when issuing host certs) and when WebPort is used
+		// in the single process mode to establish SSH reverse tunnel connection the host is validated against
+		// the valid principal list.
+		if process.Config.Proxy.WebAddr.IsHostUnspecified() {
+			port := process.Config.Proxy.WebAddr.Port(defaults.HTTPListenPort)
+			return net.JoinHostPort(string(teleport.PrincipalLocalhost), strconv.Itoa(port)), true
```

Comment:
This is really a nitpick but I think that using `PrincipalLocalhost` constant may be a little "semantically" incorrect here. I.e. that constant represents a principal encoded in the cert, not a hostname. I'd just use "localhost" verbatim here but it's up to you.

Reply:
> So that principal validation issue is only affecting "single-process" mode? Does it start ok if you launch database agent in insecure mode separately from auth/proxy?

Seems that the issue can also where reverse tunnel public address or auth address is explicitly set to `0.0.0.0` in proxy configuration

 And a traditional question, any way to cover this with a test? :)
 
 I have added DB test for this scenario.